src/constructor/constructor.py

```diff
@@ -0,0 +1,259 @@
+            if 'scheduler' in optim_params:
+                scheduler_dict = self.__create_scheduler(optimizer, optim_params.scheduler)
+                opt_sched['lr_scheduler'] = scheduler_dict
+
+            opt_sched_list.append(opt_sched)
+
+        return opt_sched_list
+
+    @staticmethod
+    def __create_optimizer(parameters: Union[Module, Tensor, Iterable[Union[Module, Tensor]]],
+                           optim_params: DictConfig) -> Optimizer:
+        optimizer_class = OPTIMIZERS.get(optim_params.optimizer.name)
+        parameters = Constructor.__set_weight_decay_for_parameters(parameters)
+        optimizer = optimizer_class(parameters, **optim_params.optimizer.params)
+
+        return optimizer
+
+    @staticmethod
+    def __create_scheduler(optimizer: Optimizer, scheduler_params: DictConfig) -> Dict[str, Any]:
+        scheduler_class = SCHEDULERS.get(scheduler_params.name)
+        scheduler = scheduler_class(optimizer, **scheduler_params.params)
+        pl_params = scheduler.pl_params if 'pl_params' in scheduler else {}
+
+        return {
+            'scheduler': scheduler,
+            **pl_params
+        }
+
+    @staticmethod
+    def __set_weight_decay_for_parameters(parameters: Union[Module, Tensor, Iterable[Union[Module, Tensor]]]
+                                          ) -> List[Dict[str, Any]]:
+        if not isinstance(parameters, Iterable) and not isinstance(parameters, Module) and \
+                not isinstance(parameters, Tensor):
+            raise ValueError(f'Unsupported parameters type for optimizer: {type(parameters)}')
+        elif not isinstance(parameters, Iterable):
+            parameters = [parameters]
+
+        param_groups = []
+        for model in parameters:
+            if isinstance(model, Module):
+                param_groups.extend(Constructor.__param_groups_weight_decay(model))
+            elif isinstance(model, Tensor) and model.requires_grad:
+                param_groups.append({'params': model})
+
+        return param_groups
+
+    # Adapted from https://github.com/rwightman/pytorch-image-models/blob/master/timm/optim/optim_factory.py
+    # Copyright 2019 Ross Wightman
+    # Licensed under The Apache 2.0 License [see LICENSE for details]
+    @staticmethod
+    def __param_groups_weight_decay(model: Module):
+        no_weight_decay_list = []
+        if hasattr(model, 'no_weight_decay'):
+            no_weight_decay_list = model.no_weight_decay()
+
+        decay = []
+        no_decay = []
+        for name, param in model.named_parameters():
+            if not param.requires_grad:
+                continue
+
+            if param.ndim <= 1 or name.endswith('.bias') or name in no_weight_decay_list:
+                no_decay.append(param)
+            else:
+                decay.append(param)
+
+        return [
+            {'params': no_decay, 'weight_decay': 0.},
+            {'params': decay}]
+
+    def create_dataloaders(self, phase: str) -> List[DataLoader]:
```

Comment:
Why return type is not Dictionary?

Reply:
PL successfully works with sequences: https://pytorch-lightning.readthedocs.io/en/1.6.2/common/lightning_module.html#train-dataloader  
And list is a sequence:
```
>>> isinstance([234], Sequence)
True
```